File: src/org/webfabric/web/sitemesh/StringTemplateDecoratorServlet.java
package org.webfabric.web.sitemesh;

import com.opensymphony.module.sitemesh.HTMLPage;
import com.opensymphony.module.sitemesh.RequestConstants;
import org.antlr.stringtemplate.StringTemplate;
import org.antlr.stringtemplate.StringTemplateGroup;
import org.antlr.stringtemplate.language.DefaultTemplateLexer;
import org.webfabric.io.Path;
import org.webfabric.web.servlet.*;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

import fj.data.List;
import static fj.data.List.list;
import fj.F;

public class StringTemplateDecoratorServlet extends HttpServlet {
    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        StringTemplate template = getTemplate(request);
        template.setAttribute("params", request.getParameterMap());
        template.setAttribute("base", ContextPath.create(request));

        HTMLPage html = getPage(request);

        StringTemplateDecorator templateDecorator = new StringTemplateDecorator(template);
        templateDecorator.Decorate(html, response.getWriter());
    }

    private StringTemplate getTemplate(final HttpServletRequest request) {
        List<Path> possiblePaths = list(OriginalServletPath.create(request),
                OriginalPathInfo.create(request),
                ServletPath.create(request),
                PathInfo.create(request));

        Path validPath = possiblePaths.dropWhile(new F<Path, Boolean>() {
            public Boolean f(Path path1) {
                return path1.value() == null;
            }
        }).head();
        String name = removeExtension(validPath.value());

        StringTemplateGroup groups = new StringTemplateGroup("decorators", getRootPath(), DefaultTemplateLexer.class);
        return groups.getInstanceOf(name);
    }

    private String getRootPath() {
        return getServletContext().getRealPath("");
    }

    private String removeExtension(String path) {
        return path.replace(".st", "");
    }

    private HTMLPage getPage(HttpServletRequest request) {
        return (HTMLPage) request.getAttribute(RequestConstants.PAGE);
    }

}

File: src/org/webfabric/web/sitemesh/StringTemplateDecoratorServlet.scala
package org.webfabric.web.sitemesh

import javax.servlet.http._
import stringtemplate._
import web.servlet._
import com.opensymphony.module.sitemesh.{HTMLPage, RequestConstants}
import io.{Url, Path}
import antlr.stringtemplate.{StringTemplate}

class StringTemplateDecoratorServlet extends HttpServlet {
  override def doPost(request: HttpServletRequest, response: HttpServletResponse) = {
    handle(request, response)
  }

  override def doGet(request: HttpServletRequest, response: HttpServletResponse) = {
    handle(request, response)
  }

  def handle(request: HttpServletRequest, response: HttpServletResponse) {
    val templateDecorator = new StringTemplateDecorator(getTemplate(request))
    templateDecorator.setBase(ContextPath(request))
    templateDecorator.setQueryString(QueryString(request))
    templateDecorator.setInclude(getPageMap(request, response))

    getPage(request) match {
      case Some(page) => templateDecorator.setPage(page)
      case None =>
    }

    templateDecorator.writeTo(response.getWriter())
  }

  def getTemplate(request: HttpServletRequest): StringTemplate = {
    val possiblePaths = List[String](OriginalServletPath(request),
      OriginalPathInfo(request),
      ServletPath(request),
      PathInfo(request))

    val StringTemplateDecoratorServlet.TemplateName(name) = possiblePaths.find(_ != null).get
    val groups = new UrlStringTemplateGroup("decorators", getBaseUrl())
    return groups.getInstanceOf(name)
  }

  def getBaseUrl(): Url = {
    getServletContext().getResource("/")
  }

  def getPage(request: HttpServletRequest): Option[HTMLPage] = {
    request.getAttribute(RequestConstants.PAGE) match {
      case page: HTMLPage => Some(page)
      case _ => None
    }
  }

  def getPageMap(request: HttpServletRequest, response: HttpServletResponse):PageMap = {
    val loaders = new PageLoaders(new UrlPageLoader, new ServletPageLoader(request, response, getServletConfig))
    new PageMap(loaders)
  }
}

object StringTemplateDecoratorServlet {
  private val TemplateName = """(.+)\.st$""".r
}